x-pack/test_serverless/shared/services/user_manager/svl_user_manager.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { REPO_ROOT } from '@kbn/repo-info';
+import { resolve } from 'path';
+import * as fs from 'fs';
+import Url from 'url';
+import { load as loadYaml } from 'js-yaml';
+import { Cookie } from 'tough-cookie';
+import { createNewSAMLSession, createSessionWithFakeSAMLAuth } from './saml_auth';
+import { FtrProviderContext } from '../../../functional/ftr_provider_context';
+
+export interface User {
+  readonly username: string;
+  readonly password: string;
+}
+
+export type Role = string;
+
+export class Session {
+  readonly cookie;
+  readonly username;
+  readonly fullname;
+  constructor(cookie: Cookie, username: string, fullname: string) {
+    this.cookie = cookie;
+    this.username = username;
+    this.fullname = fullname;
+  }
+
+  getCookieValue() {
+    return this.cookie.value;
+  }
+}
+
+export function SvlUserManagerProvider({ getService }: FtrProviderContext) {
+  const kibanaServer = getService('kibanaServer');
+  const config = getService('config');
+  const log = getService('log');
+  const isServerless = config.get('serverless');
+  const isCloud = !!process.env.TEST_CLOUD;
+  const cloudRoleUsersFilePath = resolve(REPO_ROOT, '.ftr', 'role_users.json');
+  const rolesDefinitionFilePath = resolve(
+    REPO_ROOT,
+    'packages/kbn-es/src/serverless_resources/roles.yml'
+  );
+  const roles: string[] = Object.keys(loadYaml(fs.readFileSync(rolesDefinitionFilePath, 'utf8')));
+  let roleToUserMap: Map<string, User> = new Map<string, User>();
+
+  if (!isServerless) {
+    throw new Error(`'svlUserManager' service can't be used in non-serverless FTR context`);
+  }
+
+  if (isCloud) {
+    // QAF should prepare the '.ftr/role_users.json' file for MKI pipelines
+    if (!fs.existsSync(cloudRoleUsersFilePath)) {
+      throw new Error(
+        `svlUserManager service requires user roles to be defined in ${cloudRoleUsersFilePath}`
+      );
+    }
+
+    const data = fs.readFileSync(cloudRoleUsersFilePath, 'utf8');
+    if (data.length === 0) {
+      throw new Error(`'${cloudRoleUsersFilePath}' is empty: no roles are defined`);
+    }
+    for (const [roleName, user] of Object.entries(JSON.parse(data)) as Array<[string, User]>) {
+      roleToUserMap.set(roleName, user);
+    }
+    roleToUserMap = new Map(Object.entries(JSON.parse(data)));
```

Comment:
This looks like some redundancy to me (first populating `roleToUserMap` in the loop and then assigning a new map object to it) . But maybe I'm reading this wrong? If it's intentional, could you please explain a bit?